[pipelining] Consolidate chunk counting between stage and schedule

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127946
* __->__ #127935

We used to have two backward chunk id counting systems, one at schedule level, the other at stage level.
(Which makes safety dependent on the two advancing hand-in-hand.)

This PR consolidates the counting system to the schedule side only, which would pass `mb_index` to the following stage calls:
`forward_one_chunk`
`backward_one_chunk`
`get_bwd_send_ops`
...

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

kwen2501 committed:
[pipelining] Consolidate backward chunk id counting between stage and schedule

[ghstack-poisoned]

H-Huang commented on torch/distributed/pipelining/PipelineStage.py:
```diff
@@ -543,16 +549,19 @@ def forward_one_chunk(
 
     def backward_one_chunk(
         self,
+        bwd_chunk_id: int,
```
This is kinda unrelated but:

In schedule we call the pieces of the batch, microbatches. While in stage we refer to them as chunks and here `chunk_id` instead of `mb_index`. There is some inconsistency in terms and Feng did bring up since in the pipeline stage class he sees `num_chunks` which he thought was chunks of the model. 

What should we do for that? Just wanted to open up the discussion, if we decide changes can be in another PR

H-Huang reviewed (approved):
LGTM!

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge -f "Failures are unrelated"

kwen2501 committed:
Update on "[pipelining] Consolidate backward chunk id counting between stage and schedule"


We used to have two backward chunk id counting systems, one at schedule level, the other at stage level.
(Which makes safety dependent on the two advancing hand-in-hand.)

This PR consolidates the counting system to the schedule side only, which would pass `bwd_mb_index` to the following stage calls:
`backward_one_chunk`
`get_bwd_send_ops`.

Outlier:
`get_bwd_recv_ops`, which has its own counter now: `recv_bwd_chunk_id`. I have yet to confirm why we need this and how to deal with it.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[pipelining] Consolidate chunk counting between stage and schedule"


We used to have two backward chunk id counting systems, one at schedule level, the other at stage level.
(Which makes safety dependent on the two advancing hand-in-hand.)

This PR consolidates the counting system to the schedule side only, which would pass `mb_index` to the following stage calls:
`forward_one_chunk`
`backward_one_chunk`
`get_bwd_send_ops`
...

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang reviewed (approved):
New updates LGTM

H-Huang commented:
@pytorchbot merge
